File: ms-auth/src/main/java/com/piedraazul/msauth/model/UsuarioFactory.java
package com.piedraazul.msauth.model;

// ══════════════════════════════════════════════════════
// PATRÓN GOF: FACTORY METHOD
// Propósito: centralizar la creación de usuarios según
// su rol, sin que el Service sepa los detalles de cada tipo.
// HU-06 (crear admin), HU-07 (médico), HU-08 (terapista),
// HU-09 (paciente)
// ══════════════════════════════════════════════════════
public class UsuarioFactory {

    public static Usuario crear(String username,
                                String password,
                                String nombre,
                                String email,
                                String rol,
                                Long pacienteId,
                                Long medicoId) {
        Rol rolEnum = switch (rol.toUpperCase()) {
            case "ADMINISTRADOR"    -> Rol.ADMINISTRADOR;
            case "MEDICO_TERAPISTA" -> Rol.MEDICO_TERAPISTA;
            case "AGENDADOR"        -> Rol.AGENDADOR;
            case "PACIENTE"         -> Rol.PACIENTE;
            default -> throw new IllegalArgumentException(
                    "Rol no válido: " + rol
            );
        };

        Usuario usuario = new Usuario();
        usuario.setUsername(username);
        usuario.setPassword(password);
        usuario.setNombre(nombre);
        usuario.setEmail(email);
        usuario.setRol(rolEnum);
        usuario.setActivo(true);
        usuario.setPacienteId(pacienteId);
        usuario.setMedicoId(medicoId);
        return usuario;
    }
}

File: backend/src/modules/auth/domain/usuario.factory.js
import { Usuario } from "./usuario.entity.js";
import { Rol } from "./rol.js";

export class UsuarioFactory {
  static crear({
    username,
    password,
    nombre,
    email,
    rol,
    pacienteId = null,
    medicoId = null
  }) {
    if (!Object.values(Rol).includes(rol)) {
      throw new Error(`Rol no válido: ${rol}`);
    }

    return new Usuario({
      username,
      password,
      nombre,
      email,
      rol,
      activo: true,
      pacienteId,
      medicoId
    });
  }
}